src/rime/gear/corrector.cc

```diff
@@ -0,0 +1,313 @@
+//
+// Copyright RIME Developers
+// Distributed under the BSD License
+//
+// Created by nameoverflow on 2018/11/14.
+//
+
+#include <numeric>
+#include <algorithm>
+#include <queue>
+#include <rime/service.h>
+#include <rime/ticket.h>
+#include <rime/schema.h>
+#include "rime/gear/corrector.h"
+
+using namespace rime;
+using namespace corrector;
+
+static hash_map<char, hash_set<char>> keyboard_map = {
```

Comment:
using a `vector<char>` or `string` is more economic than using `hash_set<char>` here.
The set is most of the time traversed instead of random accessed by key.

Reply:
This is designed for random access. Maybe rb-tree map?

Reply:
Oh I missed the `find()` call in `SubstCost()`.
Resolved.